### PR TITLE
GPU mode of hierarchical softmax

### LINKED ISSUE #262: Implement GPU mode for hierarchical softmax

### PR DESCRIPTION
**Pull request by unnonouno**
fix #262

### CODE REVIEW

File: tests/functions_tests/test_hierarchical_softmax.py [hunk 1]
```diff
@@ -34,19 +39,40 @@ def setUp(self):
...
+
+    @attr.gpu
+    @condition.retry(3)
+    def test_sum_gpu(self):
+        x = numpy.array([[1.0, 2.0, 3.0]], numpy.float32)
+        self.func.to_gpu()
+        self.check_sum(cuda.to_gpu(x), gpu=True)
+
+    @attr.gpu
+    def test_forwrd(self):
```
Comment (delta2323) [inline tests/functions_tests/test_hierarchical_softmax.py]: typo: forward

File: chainer/functions/hierarchical_softmax.py [hunk 2]
```diff
@@ -163,6 +164,112 @@ def _backward_cpu_one(self, x, t, gloss):
         self.gW[path] += gw
```
Comment (delta2323) [inline chainer/functions/hierarchical_softmax.py]: Do we need to use temporal variable `gw` ?

File: chainer/functions/hierarchical_softmax.py [hunk 3]
```diff
@@ -163,6 +164,112 @@ def _backward_cpu_one(self, x, t, gloss):
...
+        x, t = inputs
+        gloss, = loss
+
+        n_in = x.shape[1]
+        gx = cuda.zeros_like(x)
+        cuda.elementwise(
+            '''const float* wxy, float* gx, float* gw, const float* x,
+            const float* w, const int* ts, const int* paths,
+            const float* codes, const int* begins,
+            const float* gloss, int c, int max_length''',
```
Comment (delta2323) [inline chainer/functions/hierarchical_softmax.py]: Do we have to pass gloss as a pointer?

File: tests/functions_tests/test_hierarchical_softmax.py (hunk 1 above)
Comment (delta2323) [inline tests/functions_tests/test_hierarchical_softmax.py]: We need unit test for forward for cpu, altough that is not the focus of this PR.

File: chainer/functions/hierarchical_softmax.py (hunk 2 above)
Comment (unnonouno) [inline chainer/functions/hierarchical_softmax.py]: for readability

**Comment by unnonouno** [comment]
I fixed it about your comments.

**Comment by delta2323** [comment]
LGTM :+1:

**Pull request merged**